pkg/config/initoptions.go

```diff
@@ -0,0 +1,179 @@
+
+package config
+
+import (
+	"fmt"
+	"math/rand"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"time"
+
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	"sigs.k8s.io/kustomize/kyaml/kio"
+)
+
+const manifestFilename = "grouping-object-template.yaml"
+
+const configMapTemplate = `# NOTE: auto-generated
+# Date: <DATETIME>
+#
+# Contains the "grouping object" template ConfigMap.
+# When this object is applied, it is handled specially,
+# storing the metadata of all the other objects applied.
+# This object and its stored inventory is subsequently
+# used to calculate the set of objects to automatically
+# delete (prune), when an object is omitted from further
+# applies.
+#
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  namespace: <NAMESPACE>
+  name: grouping-object
+  labels:
+    # This label defines the ConfigMap as a grouping object template
+    cli-utils.sigs.k8s.io/inventory-id: <GROUPNAME>
+`
+
+// InitOptions contains the fields necessary to generate a
+// grouping object template ConfigMap.
+type InitOptions struct {
+	ioStreams genericclioptions.IOStreams
+	// Package directory argument
+	Dir string
+	// Namespace for grouping object
+	Namespace string
+	// Grouping object label value
+	GroupName string
+	// Random seed
+	Seed int64
+}
+
+func NewInitOptions(ioStreams genericclioptions.IOStreams) *InitOptions {
+	return &InitOptions{
+		ioStreams: ioStreams,
+	}
+}
+
+// Complete fills in the InitOptions fields.
+func (i *InitOptions) Complete(args []string) error {
+	i.Seed = time.Now().UnixNano()
+	if len(args) != 1 {
+		return fmt.Errorf("need one 'directory' arg; have %d", len(args))
+	}
+	i.Dir = args[0]
+	if !isDirectory(i.Dir) {
+		return fmt.Errorf("invalid directory argument: %s", i.Dir)
+	}
+	if len(i.Namespace) == 0 {
+		namespace, err := packageNamespace(i.Dir)
+		if err != nil {
+			return err
+		}
+		i.Namespace = namespace
+	}
+	if len(i.GroupName) == 0 {
+		i.GroupName = i.defaultGroupName()
+	}
+	if !validateGroupName(i.GroupName) {
+		return fmt.Errorf("invalid group name: %s", i.GroupName)
+	}
+	return nil
+}
+
+// isDirectory returns true if the passed path is a directory;
+// false otherwise.
+func isDirectory(path string) bool {
+	if d, err := os.Stat(path); err == nil {
+		if d.IsDir() {
+			return true
+		}
+	}
+	return false
+}
+
+// packageNamespace returns the namespace of the package
+// config files.
+func packageNamespace(packageDir string) (string, error) {
+	r := kio.LocalPackageReader{PackagePath: packageDir}
+	nodes, err := r.Read()
+	if err != nil {
+		return "", err
+	}
+	namespace := "default"
+	for _, node := range nodes {
+		rm, err := node.GetMeta()
+		if err != nil {
+			continue
+		}
+		if len(rm.ObjectMeta.Namespace) > 0 {
+			namespace = rm.ObjectMeta.Namespace
+		}
+		break
+	}
+	return namespace, nil
+}
+
+// defaultGroupName returns a string of the package name (directory)
+// with a random number suffix.
+func (i *InitOptions) defaultGroupName() string {
```

Comment:
Do we need the random suffix? I think it would be sufficient to just use the name of the package. If that isn't unique or not valid, we can just let the user decide on a name.

Reply:
Phil asked for it. We name the groupname after the package directory name, And he thought that if the package is under a directory called "prod", then there might be several of them, and we might need the suffix to disambiguate. I'm happy to remove. Adding a TODO.

Reply:
Yeah, I guess it doesn't really matter if the name has any meaning or not as long as it is unique and it is checked in with the rest of the manifests.